File: recoverer.js
const fs = require("fs").promises;
const path = require("path");
const process = require("process");
const { authenticate } = require("@google-cloud/local-auth");
const { google } = require("googleapis");

// If modifying these scopes, delete token.json.
const SCOPES = ["https://www.googleapis.com/auth/drive"];
// The file token.json stores the user's access and refresh tokens, and is
// created automatically when the authorization flow completes for the first
// time.
const TOKEN_PATH = path.join(process.cwd(), "token.json");
const CREDENTIALS_PATH = path.join(process.cwd(), "credentials.json");

const PICTURES_FOLDER = "1X9YxmaM1eFim921bN_8BbJbheQP99eMu";

/**
 * Reads previously authorized credentials from the save file.
 *
 * @return {Promise<OAuth2Client|null>}
 */
async function loadSavedCredentialsIfExist() {
	try {
		const content = await fs.readFile(TOKEN_PATH);
		const credentials = JSON.parse(content);
		return google.auth.fromJSON(credentials);
	} catch (err) {
		return null;
	}
}

/**
 * Serializes credentials to a file comptible with GoogleAUth.fromJSON.
 *
 * @param {OAuth2Client} client
 * @return {Promise<void>}
 */
async function saveCredentials(client) {
	const content = await fs.readFile(CREDENTIALS_PATH);
	const keys = JSON.parse(content);
	const key = keys.installed || keys.web;
	const payload = JSON.stringify({
		type: "authorized_user",
		client_id: key.client_id,
		client_secret: key.client_secret,
		refresh_token: client.credentials.refresh_token,
	});
	await fs.writeFile(TOKEN_PATH, payload);
}

/**
 * Load or request or authorization to call APIs.
 *
 */
async function authorize() {
	let client = await loadSavedCredentialsIfExist();
	if (client) {
		return client;
	}
	client = await authenticate({
		scopes: SCOPES,
		keyfilePath: CREDENTIALS_PATH,
	});
	if (client.credentials) {
		await saveCredentials(client);
	}
	return client;
}

const TARGET_FOLDER = "10anu-USkcZwx9heqUm-IYOkRjNNYsnUh";
let foldersToCheck = [TARGET_FOLDER];
const allFoldersFound = [TARGET_FOLDER];
const allFilesFound = [];
let drive;

async function listFolders(target, nextPageToken) {
	const res = await drive.files.list({
		fields: "nextPageToken, files(id, name)",
		pageSize: 3,
		q: `'${target}' in parents and mimeType = 'application/vnd.google-apps.folder'`,
		...(nextPageToken && { pageToken: nextPageToken }),
	});

	const folders = res.data.files;
	if (folders.length === 0) {
		console.log("No folders found.");
		return;
	}

	console.log("Folders:");
	foldersToCheck.push(
		...folders.map(folder => {
			console.log(`${folder.name}`);
			allFoldersFound.push(folder.id);
			return folder.id;
		})
	);

	if (res.data.nextPageToken) {
		listFolders(target, res.data.nextPageToken);
		return;
	}
	while (foldersToCheck.length !== 0) {
		listFolders(foldersToCheck.shift());
	}
}

async function listFiles(target, nextPageToken) {
	const res = await drive.files.list({
		fields: "nextPageToken, files(id, name)",
		pageSize: 15,
		q: `'${target}' in parents and (mimeType contains 'image/' or mimeType contains 'video/')`,
		...(nextPageToken && { pageToken: nextPageToken }),
	});

	const files = res.data.files;
	if (files.length === 0) {
		console.log("No files found.");
		return;
	}

	console.log("Files:");
	files.forEach(file => {
		console.log(`${file.name}`);
		allFilesFound.push({ id: file.id, name: `${target}/${file.name}` });
	});

	if (res.data.nextPageToken) {
		listFiles(target, res.data.nextPageToken);
		return;
	}
	while (foldersToCheck.length !== 0) {
		listFiles(foldersToCheck.shift());
	}
}

async function getOldestRevision({ id: fileId, name: fileName }, nextPageToken) {
	const res = await drive.revisions.list({
		fileId: fileId,
		pageSize: 10,
		...(nextPageToken && { pageToken: nextPageToken }),
	});
	const revisions = res.data.revisions;
	if (revisions.length === 0) {
		console.log("No revisions found?");
		return;
	}

	console.log(`Revisions for ${fileName}:`);
	revisions.forEach(rev => {
		console.log(`${rev.modifiedTime}`);
	});

	if (res.data.nextPageToken) {
		getOldestRevision(target, res.data.nextPageToken);
		return;
	}

	const oldestRevId = res.data.revisions[res.data.revisions.length - 1].id;
	const binFileResponse = await drive.revisions.get({
		fileId: fileId,
		revisionId: oldestRevId,
	});
	await fs.writeFile(fileName, binFileResponse.data);
}

authorize()
	.then(async authClient => {
		drive = google.drive({ version: "v3", auth: authClient });
		// await listFolders(authClient, foldersToCheck.shift());
		foldersToCheck = []; // allFoldersFound;
		await listFiles("1VoOxjStf8_R8ILuz4snNjLbVlYFIKHh2");
		getOldestRevision(allFilesFound[0]);
		// await listFiles(authClient, foldersToCheck.shift())
	})
	.catch(console.error);

File: recoverer.py
from __future__ import print_function
import io
import shutil

import os.path

from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError
from googleapiclient.http import MediaIoBaseDownload

# If modifying these scopes, delete the file token.json.
SCOPES = ['https://www.googleapis.com/auth/drive']

service = None
creds = None
allFoldersFound = []
foldersToCheck = ['10anu-USkcZwx9heqUm-IYOkRjNNYsnUh']
allFilesFound = []


def listFolders(target, nextPageToken=None, prevRequest=None):
    global service, allFoldersFound, foldersToCheck
    if nextPageToken != None and prevRequest != None:
        request = service.files().list_next(
            prevRequest, {"nextPageToken": nextPageToken})
    else:
        request = service.files().list(
            pageSize=10,
            fields="nextPageToken, files(id, name)",
            q=f"'{target}' in parents and mimeType = 'application/vnd.google-apps.folder'",
        )
    results = request.execute()
    folders = results.get('files', [])

    if not folders:
        print('No folders found.')
        return
    print('Folders:')
    for folder in folders:
        print(u'{0} ({1})'.format(folder['name'], folder['id']))
        path = f"{os.curdir}{os.sep}downloads{os.sep}{folder['name']}"
        if not os.path.exists(path):
            os.makedirs(path)
        allFoldersFound.append({'id':folder['id'], 'path': path})
        foldersToCheck.append(folder['id'])

    nextPageToken = results.get('nextPageToken', None)
    if nextPageToken is not None:
        listFolders(target, nextPageToken, request)
        return
    while len(foldersToCheck) > 0:
        listFolders(foldersToCheck.pop())


def listFiles(target, nextPageToken=None, prevRequest=None):
    global service, allFilesFound

    if nextPageToken != None and prevRequest != None:
        request = service.files().list_next(
            prevRequest, {"nextPageToken": nextPageToken})
    else:
        request = service.files().list(
            pageSize=10,
            fields="nextPageToken, files(id, name)",
            q=f"'{target['id']}' in parents and (mimeType contains 'image/' or mimeType contains 'video/')",
        )
    results = request.execute()
    files = results.get('files', [])

    if not files:
        print('No files found.')
        return
    print('Files:')
    for file in files:
        print(u'{0} ({1})'.format(file['name'], file['id']))
        allFilesFound.append(
            {"id": file['id'], "name": f"{target['path']}{os.sep}{file['name']}"})

    nextPageToken = results.get('nextPageToken', None)
    if nextPageToken is not None:
        listFiles(target, nextPageToken, request)
        return

def getOldestRevision(file, nextPageToken=None, prevRequest=None):
    global service

    if nextPageToken != None and prevRequest != None:
        request = service.revisions().list_next(
            prevRequest, {"nextPageToken": nextPageToken})
    else:
        request = service.revisions().list(
            pageSize=10,
            fileId=file['id']
        )
    results = request.execute()
    revisions: list = results.get('revisions', [])

    if not revisions:
        print('No revisions found?')
        return
    print(f"Revisions for file {file['name']}:")
    for rev in revisions:
        print(f"{rev['modifiedTime']}")

    nextPageToken = results.get('nextPageToken', None)
    if nextPageToken is not None:
        getOldestRevision(file, nextPageToken, request)
        return
 
    oldestRevisionId = revisions.pop()['id']
    downloadFile(file, oldestRevisionId)


def downloadFile(file, revisionId):
    serviceV2 = build('drive', 'v2', credentials=creds)
    # request = service.files().get_media(fileId=file['id'])
    request = serviceV2.revisions().get(fileId=file['id'], revisionId=revisionId)
    res = request.execute()
    request.uri = res.get('downloadUrl')
    # fh = io.FileIO(f"{file['name']}", mode='w+b')
    fh = io.BytesIO()
    downloader = MediaIoBaseDownload(fh, request)

    done = False
    while done is False:
        status, done = downloader.next_chunk()
        if status:
            print("Download %d%%." % int(status.progress() * 100))
    with io.open(file['name'], "wb") as f:
        fh.seek(0)
        f.write(fh.read())


def main():
    global creds
    """Shows basic usage of the Drive v3 API.
    Prints the names and ids of the first 10 files the user has access to.
    """
    creds = None
    # The file token.json stores the user's access and refresh tokens, and is
    # created automatically when the authorization flow completes for the first
    # time.
    if os.path.exists('token.json'):
        creds = Credentials.from_authorized_user_file('token.json', SCOPES)
    # If there are no (valid) credentials available, let the user log in.
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file(
                'credentials.json', SCOPES)
            creds = flow.run_local_server(port=0)
        # Save the credentials for the next run
        with open('token.json', 'w') as token:
            token.write(creds.to_json())

    try:
        global service, allFilesFound, allFoldersFound
        service = build('drive', 'v3', credentials=creds)
        listFolders(foldersToCheck.pop())
        for folder in allFoldersFound:
            listFiles(folder)
        for file in allFilesFound:
            getOldestRevision(file)
    

        # await listFiles(authClient, foldersToCheck.shift())
    except HttpError as error:
        # TODO(developer) - Handle errors from drive API.
        print(f'An error occurred: {error}')


if __name__ == '__main__':
    main()
